In [49]:
import numpy as numpy
import pandas as pd
import matplotlib as plt 
import seaborn as sns
from utils import load_data
plt.style.use('ggplot')
pd.set_option("display.max_columns", None)

In [50]:

df = load_data()
df.shape

Using cached files.


(112096, 39)

In [51]:
df.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date',
       'price', 'freight_value', 'product_category_name',
       'product_name_lenght', 'product_description_lenght',
       'product_photos_qty', 'product_weight_g', 'product_length_cm',
       'product_height_cm', 'product_width_cm', 'seller_zip_code_prefix',
       'seller_city', 'seller_state', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'seller_lat', 'seller_lng', 'customer_lat', 'customer_lng',
       'payment_installments', 'payment_type', 'payment_value',
       'payment_sequential', 'review_score', 'review_comment_length'],
      dtype='str')

Target Metric (actual_delivery_days)

In [ ]:
df["actual_delivary_days"] = (df["order_delivered_customer_date"] - df["order_purchase_timestamp"]).dt.days
df = df.dropna(subset=["actual_delivery_days"])

y = df["actual_delivery_days"]
y.describe()

TypeError: dtype timedelta64[ms] cannot be converted to datetime64[ns]

In [ ]:
df.dtypes

order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[ms]
order_approved_at                datetime64[ms]
order_delivered_carrier_date     datetime64[ms]
order_delivered_customer_date    datetime64[ms]
order_estimated_delivery_date    datetime64[ms]
order_item_id                             int64
product_id                                  str
seller_id                                   str
shipping_limit_date              datetime64[ms]
price                                   float64
freight_value                           float64
product_category_name                       str
product_name_lenght                     float64
product_description_lenght              float64
product_photos_qty                      float64
product_weight_g                        float64
product_length_cm                       float64
product_height_cm                       

In [ ]:
df.describe()

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,order_item_id,shipping_limit_date,price,freight_value,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,seller_zip_code_prefix,customer_zip_code_prefix,seller_lat,seller_lng,customer_lat,customer_lng,payment_installments,payment_value,payment_sequential,review_score,review_comment_length
count,112096,112081,110913,109660,112096,112096.000000,112096,112096.000000,112096.000000,110500.000000,110500.000000,110500.000000,112078.000000,112078.000000,112078.000000,112078.000000,112096.000000,112096.000000,112096.000000,112096.000000,112096.000000,112096.000000,112093.000000,112093.000000,112093.000000,111156.000000,47247.000000
mean,2018-01-01 00:49:34.837000,2018-01-01 12:26:22.950000,2018-01-05 03:00:07.901000,2018-01-14 13:43:19.531000,2018-01-24 20:45:49.842000,1.198036,2018-01-07 16:04:26.382000,120.682327,19.982875,48.763566,787.953457,2.204561,2092.227315,30.145773,16.613421,22.986786,24446.856293,35041.234629,-22.798601,-47.236316,-21.232442,-46.201583,3.017111,180.324445,1.044829,4.033131,70.268931
min,2016-09-04 21:15:19,2016-09-15 12:16:38,2016-10-08 10:34:01,2016-10-11 13:46:32,2016-10-04 00:00:00,1.000000,2016-09-19 00:15:34,0.850000,0.000000,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000,1001.000000,1003.000000,-32.079231,-67.813284,-33.689948,-72.668881,0.000000,9.590000,1.000000,1.000000,1.000000
25%,2017-09-13 19:20:49.750000,2017-09-14 03:06:08,2017-09-18 20:54:33,2017-09-26 20:09:53.500000,2017-10-05 00:00:00,1.000000,2017-09-20 21:01:58.500000,39.900000,13.070000,42.000000,348.000000,1.000000,300.000000,18.000000,8.000000,15.000000,6429.000000,11095.000000,-23.607473,-48.829744,-23.590817,-48.112804,1.000000,65.560000,1.000000,4.000000,29.000000
50%,2018-01-20 01:06:48,2018-01-20 15:41:14,2018-01-24 19:16:33,2018-02-02 21:25:32.500000,2018-02-16 00:00:00,1.000000,2018-01-26 14:03:07.500000,74.900000,16.250000,52.000000,602.000000,1.000000,700.000000,25.000000,13.000000,20.000000,13568.000000,24240.000000,-23.425556,-46.744092,-22.929384,-46.633932,2.000000,114.420000,1.000000,5.000000,54.000000
75%,2018-05-04 19:08:13.250000,2018-05-05 03:33:40,2018-05-08 13:40:00,2018-05-15 20:37:09.750000,2018-05-28 00:00:00,1.000000,2018-05-10 15:10:22,134.900000,21.150000,57.000000,986.000000,3.000000,1800.000000,38.000000,20.000000,30.000000,27700.000000,58430.000000,-21.757321,-46.518679,-20.171285,-43.635297,4.000000,195.400000,1.000000,5.000000,98.000000
max,2018-09-03 09:06:57,2018-09-03 17:40:06,2018-09-11 19:48:28,2018-10-17 13:22:46,2018-10-25 00:00:00,21.000000,2020-04-09 22:35:08,6735.000000,409.680000,76.000000,3992.000000,20.000000,40425.000000,105.000000,105.000000,118.000000,99730.000000,99990.000000,-2.501242,-34.855616,42.184003,-8.723762,24.000000,13664.080000,29.000000,5.000000,208.000000
std,NaN,NaN,NaN,NaN,NaN,0.705953,NaN,183.834080,15.819910,10.029791,652.740638,1.716798,3746.852629,16.145288,13.453462,11.681725,27602.765262,29853.844213,2.696503,2.339947,5.571588,4.037601,2.801854,273.198926,0.386121,1.386158,54.499388


Dropping Irrelevant Columns 

## Findings — target drift over time (investigate tomorrow)

**The target (`actual_delivery_days`) is non-stationary — it shifts strongly over time.**

- Time-based 80/20 split: **train mean = 12.9 days, test mean = 8.2 days** (a 4.7-day gap; train is the older 80%, test is the most recent 20%).
- Three eras in the monthly average (see the "Avg delivery days by month" chart):
  - **Late 2016:** ~50 days then crashing — Olist launch; very few, erratic orders.
  - **2017:** steady ~12–14 days.
  - **2018:** declining to ~7–8 days — deliveries got faster over time.
- This drift is the cause of the model R² = −2.5 on the held-out test set: it trains on the slow era and is graded on the fast tail, so it over-predicts and scores worse than guessing the mean.
- **Likely right-censoring at the tail:** the dataset ends ~Sept 2018. Orders near the end that took a long time had not been delivered at export, so they have no delivery date and were removed by `dropna`. The latest period is therefore biased toward fast deliveries — part of the 8.2 test mean is artificial, not a real improvement.
